umbraco/Umbraco-CMS
# Remove content from public facing Examine indexes when it is made protected

### Prerequisites


### Description

The public facing Examine indexes (i.e. `ExternalIndex`) filters out protected content when indexing on publish and when rebuilding  (it happens [here](https://github.com/umbraco/Umbraco-CMS/blob/release/12.0/src/Umbraco.Infrastructure/Examine/ContentValueSetValidator.cs#L85)). However, if content is made protected subsequently to a publish or index rebuild, the content remains in the indexes.

This PR ensures that content is removed from the applicable indexes when public access protection is applied.

### Limitations

We have to leverage the public access cache refreshing notification to implement this, because we need to perform this task on every single server in a load balanced setup. Unfortunately, these notifications have been implemented rather sparsely, which limits what we can do with them.

At this time we can only _remove_ content from the indexes when public access is _added_ to content. We can't _add_ content to the indexes when public access is _removed_ from content. 

To overcome this limitations, editors must republish the entire structure of the previously protected content after removing public access restrictions.

### Testing this PR

You should probably start by rebuilding the `ExternalIndex`, to ensure you've got an accurate starting point.

You also need a member group (members optional) to perform public access configuration.

1. Create a structure of published content (i.e. a parent and a child)
2. Verify that the `ExternalIndex` contains all the content.
3. Apply public access restrictions at the top level of the structure.
4. Verify that the `ExternalIndex` no longer contains any item in the structure.
5. Remove the public access restrictions again.
6. Republish the entire structure (publish with descendants at top level).
7. Verify that the `ExternalIndex` once more contains all the content.
